crud/common/rebalance.lua

```diff
@@ -0,0 +1,211 @@
+            (op == 'REPLACE' and new.status == vshard_consts.BUCKET.SENDING) then
+        box.broadcast(SAFE_MOD_ENABLE_EVENT, true)
+    end
+end
+
+local function register_enable_hook(func)
+    M.safe_mode_enable_hooks[func] = true
+end
+
+local function remove_enable_hook(func)
+    M.safe_mode_enable_hooks[func] = nil
+end
+
+local function register_disable_hook(func)
+    M.safe_mode_disable_hooks[func] = true
+end
+
+local function remove_disable_hook(func)
+    M.safe_mode_disable_hooks[func] = nil
+end
+
+local function safe_mode_status()
+    return M.safe_mode
+end
+
+local function safe_mode_enable()
+    if not box.info.ro then
+        box.space[SETTINGS_SPACE_NAME]:replace{ 'safe_mode', true }
+        for _, trig in pairs(box.space._bucket:on_replace()) do
+            if trig == safe_mode_trigger then
+                box.space._bucket:on_replace(nil, safe_mode_trigger)
+            end
+        end
+    end
+    M.safe_mode = true
+
+    for hook, _ in pairs(M.safe_mode_enable_hooks) do
+        hook()
+    end
+
+    log.info('Rebalance safe mode enabled')
+end
+
+local function safe_mode_disable()
+    if not box.info.ro then
+        box.space[SETTINGS_SPACE_NAME]:replace{ 'safe_mode', false }
+        box.space._bucket:on_replace(safe_mode_trigger)
+    end
+    M.safe_mode = false
+
+    for hook, _ in pairs(M.safe_mode_disable_hooks) do
+        hook()
+    end
+
+    log.info('Rebalance safe mode disabled')
+end
+
+local function rebalance_init()
+    M.metrics.enable_storage_metrics()
+
+    -- box.watch was introduced in tarantool 2.10.0
+    if not utils.tarantool_supports_box_watch() then
+        log.warn('This version of tarantool does not support autoswitch to safe mode during rebalance. '
+            .. 'Update to newer version or use `_crud.rebalance_safe_mode_enable()` to enable safe mode manually.')
+        return
+    end
+
+    box.watch('box.status', function()
+        if box.info.ro then
+            return
+        end
+
+        local stored_safe_mode
+        if box.space[SETTINGS_SPACE_NAME] == nil then
+            create_space()
+            box.space[SETTINGS_SPACE_NAME]:insert{ 'safe_mode', false }
+        else
+            stored_safe_mode = box.space[SETTINGS_SPACE_NAME]:get{ 'safe_mode' }
+        end
+        M.safe_mode = stored_safe_mode and stored_safe_mode.value or false
+
+        if M.safe_mode then
+            for hook, _ in pairs(M.safe_mode_enable_hooks) do
+                hook()
+            end
+        else
+            box.space._bucket:on_replace(safe_mode_trigger)
+            for hook, _ in pairs(M.safe_mode_disable_hooks) do
+                hook()
+            end
+        end
+    end)
+
+    box.watch(SAFE_MOD_ENABLE_EVENT, function(_, do_enable)
+        if box.info.ro or not do_enable then
+            return
+        end
+        safe_mode_enable()
+    end)
+end
+
+local function router_cache_clear()
```

Comment:
Aren't we supposed to check, that rebalancing is ended and we have no buckets in forbidden states?